File: nodejs/wgtrf.js
const {exec} = require("child_process");
const dgram = require("dgram");
const udpServer = dgram.createSocket("udp4");
const options = {port: (process.argv[4] || 33055), address: (process.argv[3] || "0.0.0.0")}

const cmd_show = "wg show " + (process.argv[2] || "wg0");


function parseConf(conf){
	let result={}, tmp;
	let lines = conf.split("\n");
	for (const line of lines) {
		if (line.startsWith("peer: ")){
			tmp = {"public key" : line.substr(6)};
		}else if (!tmp){
			continue;
		}else if (line.startsWith("  endpoint: ")){
			tmp["endpoint"] = line.substr(12);
		}else if (line.startsWith("  allowed ips: ")){
			tmp["allowed ips"] = line.substr(15).replace(/ /g, "");
		}else if (line === ""){
			let ip = tmp["allowed ips"];
			result[ip.substring(0, ip.indexOf("/"))] = tmp;
		}
	}
	return result;
}

// pingXXX  peerIP -> (pingXXX-ip-port reqIP -> pingXXX-ip-port pong) -> pingXXX pong
// setxXXX  peerIP -> (setxXXX-ip-port publicKey... -> setxXXX-ip-port success) ->setxXXX success reqIP
udpServer.on("message", function(msg, rinfo) {
	const cmd = msg.toString();
	console.log(new Date().getTime(), `udp recv(${rinfo.address}:${rinfo.port}):`, cmd);
	
	const cmd_ps = cmd.split(" ", 2);
	const id_ps = cmd_ps[0].split("-");
	if (id_ps.length === 1){ // front
		const peerIP = cmd_ps[1];
		if (cmd_ps[0].startsWith("ping")){
			const cmd2 = `${cmd_ps[0]}-${rinfo.address}-${rinfo.port} ${peerIP}`;
			console.log(new Date().getTime(), `udp send(${peerIP}:${options.port}):`, cmd2);
			udpServer.send(cmd2, options.port, peerIP);		
		}else{
			// 执行查询指令
			exec(cmd_show, (error, stdout, stderr) => {
				//console.log(new Date().getTime(), `stdout: ${stdout}`);
				console.error(new Date().getTime(), `stderr: ${stderr}`);
				if (error) {
					console.error(`exec error: ${error}`);
					udpServer.send(cmd_ps[0] + " fail " + error, rinfo.port,rinfo.address);
				}else{
					let conf = parseConf(stdout);
					let own = conf[rinfo.address];
					let peer = conf[peerIP];
					if (own && peer){
						let msg1 = `${cmd_ps[0]}-${rinfo.address}-${rinfo.port} ` + own["public key"];
						let msg2 = cmd_ps[0] + " " + peer["public key"];
						if (cmd_ps[0].startsWith("remove")){
							msg1 += " remove";
							msg2 += " remove";
						}else{
							msg1 += (" endpoint " + own["endpoint"] + " persistent-keepalive 25 allowed-ips " + own["allowed ips"]);
							msg2 += (" endpoint " + peer["endpoint"] + " persistent-keepalive 25 allowed-ips " + peer["allowed ips"]);
						}
						
						console.log(new Date().getTime(), `udp send(${peerIP}:${options.port}):`, msg1);
						udpServer.send(msg1, options.port, peerIP);
						
						console.log(new Date().getTime(), `udp send(${rinfo.address}:${options.port}):`, msg2);
						udpServer.send(msg2, rinfo.port, rinfo.address);
					}
				}			
			});			
		}			
	}else{ // end
		const cmd2 = `${id_ps[0]}-${rinfo.address}-${rinfo.port} ${cmd_ps[1]} ${id_ps[1]}`;
		console.log(new Date().getTime(), `udp send(${id_ps[1]}:${id_ps[2]}):`, cmd2);
		udpServer.send(cmd2, parseInt(id_ps[2]), id_ps[1]);	
	}
});

udpServer.bind(options);
console.log(new Date().getTime(), `udp server(${options.address}:${options.port}) start...`);

File: lua/wgtrf.lua

local socket = require("socket.core")

local function parseConf(conf)
	local result={}, tmp
	string.gsub(conf, "[^\n]+", function(line)
		if (line:find("^peer: ")) then
			if tmp then
				local ip = tmp["allowed ips"]
				result[ip:sub(1, ip:find("/")-1)] = tmp				
			end
			tmp = {["public key"] = line:sub(1+6)}
		elseif (line:find("^  endpoint: ")) then
			tmp["endpoint"] = line:sub(1+12)
		elseif (line:find("^  allowed ips: ")) then
			tmp["allowed ips"] = line:sub(1+15):gsub(" ", "")
		end		
	end)
	if tmp then
		local ip = tmp["allowed ips"]:match("([^/]+)")
		result[ip] = tmp				
	end
	return result
end

-- pingXXX  peerIP -> (pingXXX-ip-port reqIP -> pingXXX-ip-port pong) -> pingXXX pong
-- setxXXX  peerIP -> (setxXXX-ip-port publicKey... -> setxXXX-ip-port success) ->setxXXX success reqIP
local function start(device, address, port)
	print(string.format("[%d][wgtrf] open(address=%s, port=%d, device=%s)", os.millisecond(), address, port, device))
	
	local cmd_show = "wg show " .. device
	local m = socket.udp()
	assert(m:setsockname(address, port))
		
	while(true) do
		local cmd,remoteAddress,remotePort = m:receivefrom()
		if (not cmd) then
			print(string.format("[%d][wgtrf] err", os.millisecond()), remoteAddress)
			break
		end
		print(string.format("[%d][wgtrf] receivefrom(%s:%d), data:", os.millisecond(), remoteAddress, remotePort), cmd)
	
		local cmd_id,cmd_data = cmd:match("^([^ ]+) (.+)$")
		if (cmd_id and cmd_data) then
			local id_pos = cmd_id:find("-")
			if (not id_pos) then -- front
				local peerIP = cmd_data
				if (cmd_id:find("^ping")) then
					local cmd2 = string.format("%s-%s-%d %s", cmd_id, remoteAddress, remotePort, remoteAddress)
					print(string.format("[%d][wgtrf] sendto(%s:%d), data:", os.millisecond(), peerIP, port), cmd2)
					assert(m:sendto(cmd2, peerIP, port))
				else
					--执行查询指令
					local p = io.popen(cmd_show, "r")
					local stdout = p:read("*a")
					p:close()

					local conf = parseConf(stdout)
					local own = conf[remoteAddress]
					local peer = conf[peerIP]
					if (own and peer) then
						local msg1 = string.format("%s-%s-%d ", cmd_id, remoteAddress, remotePort) .. own["public key"]
						local msg2 = cmd_id .. " " .. peer["public key"]
						if (cmd_id == "remove") then
							msg1 = msg1 .. " remove"
							msg2 = msg2 .. " remove"
						else
							msg1 = msg1 .. (" endpoint " .. own["endpoint"] .. " persistent-keepalive 25 allowed-ips " .. own["allowed ips"])
							msg2 = msg2 .. (" endpoint " .. peer["endpoint"] .. " persistent-keepalive 25 allowed-ips " .. peer["allowed ips"])
						end
						print(string.format("[%d][wgtrf] sendto(%s:%d), data:", os.millisecond(), peerIP, port), msg1)
						assert(m:sendto(msg1, peerIP, port))
						print(string.format("[%d][wgtrf] sendto(%s:%d), data:", os.millisecond(), remoteAddress, remotePort), msg2)
						assert(m:sendto(msg2, remoteAddress, remotePort))
					end					
				end
			else -- end
				local address2,port2 = cmd_id:match("([^%-]+)%-(.+)$", id_pos)
				local cmd2 = string.format("%s-%s-%d %s %s", cmd_id:sub(1,id_pos-1), remoteAddress, remotePort, cmd_data, address2)
				print(string.format("[%d][wgtrf] sendto(%s:%s), data:", os.millisecond(), address2, port2), cmd2)
				assert(m:sendto(cmd2, address2, tonumber(port2)))
			end
		end	--if
	end --while

	m:close()

	print(string.format("[%d][wgtrf] close", os.millisecond()))
end

local function daemon(arg1, arg2, arg3, arg4)
	while(true) do
		print(string.format("[%d][wgtrf] daemon", os.millisecond()))
		local ok,err = pcall(start, arg1, arg2, arg3, arg4)
		if (not ok) then
			print(string.format("[%d][wgtrf] err", os.millisecond))
		end
		os.sleep(15)
	end
end


--arg[1] = "daemon"
if (arg[1] == "daemon") then
	daemon(arg[2], arg[3], arg[4])
else
	start(arg[1] or "wg0", arg[2] or "0.0.0.0", arg[3] or 33055)
end
